File: java/lesson 8/DfsMatrix.java
class DfsMatrix {
    private static int[][] dir = { {0, 1}, {1, 0}, {0, -1}, {-1, 0} };
    
    public void dfs(int[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        dfs(grid, 0, 0, visited);
    }
    
    private void dfs(int[][] grid, int row, int col, boolean[][] visited) {
        if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || visited[row][col]) { return; }
        
        visited[row][col] = true;        
        System.out.println(grid[row][col]);
        
        for (int i = 0; i < dir.length; i++) {
            dfs(grid, row + dir[i][0], col + dir[i][1], visited);
        }
    }
}

File: kotlin/lesson 8/DfsMatrix.kt
class DfsMatrix {
    private val dir: Array<IntArray> =  arrayOf(intArrayOf(0, 1), intArrayOf(1,0), intArrayOf(0, -1), intArrayOf(-1, 0))
    
    fun dfs(grid: Array<IntArray>) {
        val visited = Array(grid.size) { BooleanArray(grid[0].size) }
        dfs(grid, 0, 0, visited)
    }
    
    private fun dfs(grid: Array<IntArray>, row: Int, col: Int, visited: Array<BooleanArray>) {
        if (row < 0 || row >= grid.size || col < 0 || col >= grid[0].size || visited[row][col]) { return }
        
        visited[row][col] = true        
        System.out.println(grid[row][col])
        
        for (i in 0..dir.size-1) {
            dfs(grid, row + dir[i][0], col + dir[i][1], visited)
        }
    }
}
